llvm/lib/Target/AIE/AIEWawRegRewriter.cpp

```diff
@@ -0,0 +1,381 @@
+//===----- AIEWawRegRewriter.cpp - Rewrite Regs to remove Defs ---------===//
+//
+// This file is licensed under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+// (c) Copyright 2024 Advanced Micro Devices, Inc. or its affiliates
+//
+//===----------------------------------------------------------------------===//
+//
+// Implementation of a Loop Register rewriting pass, that reduces WAW physical
+// register conflicts, so that software pipelining has more freedom to operate.
+//
+//===----------------------------------------------------------------------===//
+
+#include "AIEBaseInstrInfo.h"
+#include "AIEBaseRegisterInfo.h"
+#include "Utils/AIELoopUtils.h"
+
+#include "llvm/ADT/SmallVector.h"
+#include "llvm/CodeGen/LiveDebugVariables.h"
+#include "llvm/CodeGen/LiveIntervals.h"
+#include "llvm/CodeGen/LiveRegMatrix.h"
+#include "llvm/CodeGen/LiveStacks.h"
+#include "llvm/CodeGen/MachineFunctionPass.h"
+#include "llvm/CodeGen/MachineInstr.h"
+#include "llvm/CodeGen/MachineOperand.h"
+#include "llvm/CodeGen/Passes.h"
+#include "llvm/CodeGen/SlotIndexes.h"
+#include "llvm/CodeGen/TargetInstrInfo.h"
+#include "llvm/CodeGen/TargetSubtargetInfo.h"
+#include "llvm/MC/MCRegister.h"
+#include "llvm/Support/Debug.h"
+#include "llvm/Support/raw_ostream.h"
+#include <llvm/CodeGen/MachineBasicBlock.h>
+
+using namespace llvm;
+
+#define DEBUG_TYPE "aie-waw-reg-rewrite"
+
+namespace {
+
+///
+/// This rewrites Registers, if they are redefined in the same Loop body and
+/// could cause WAW issues in the loop pipelining.
```

Comment:
Super nit: I would maybe rephrase to make the message a bit more general, and less prone to getting outdated. E.g. `This rewrites physical register assignments in critical parts of the code (like loops) to break WAW and WAR dependencies.`